[DataPipe] refactor to separate `_IterDataPipeMeta`

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70216
* #70373
* #76999
* #75995
* #70479
* __->__ #76776

There are some features of the metaclass that we would like to be unique to `IterDataPipe` and not applied to `MapDataPipe`. 

For example, `hook_iterator` should only impact `__iter__` of `IterDataPipe` since `MapDataPipe` has random access to the data and should not be constrained.

**Note that I plan to land this together the single iterator constraint PRs, so this will not land for now.**

Differential Revision: [D36156459](https://our.internmc.facebook.com/intern/diff/D36156459)

NivekT committed:
[DataPipe] refactor to separate

[ghstack-poisoned]

ejguan reviewed (approved):
LGTM, thanks a lot!

NivekT committed:
Update on "[DataPipe] refactor to separate `_IterDataPipeMeta`"


There are some features of the metaclass that we would like to be unique to `IterDataPipe` and not applied to `MapDataPipe`. 

For example, `hook_iterator` should only impact `__iter__` of `IterDataPipe` since `MapDataPipe` has random access to the data and should not be constrained.

**Note that I plan to land this together the single iterator constraint PRs, so this will not land for now.**

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36156459).

ejguan commented on torch/utils/data/datapipes/_typing.py:
```diff
@@ -253,11 +253,9 @@ class _DataPipeMeta(GenericMeta):
-
         return super().__new__(cls, name, bases, namespace, **kwargs)  # type: ignore[call-overload]
 
+        # TODO: the statements below are not reachable by design as there is a bug and typing is low priority for now.
```
Thanks for the comment for clarification!

NivekT committed:
Update on "[DataPipe] refactor to separate `_IterDataPipeMeta`"


There are some features of the metaclass that we would like to be unique to `IterDataPipe` and not applied to `MapDataPipe`. 

For example, `hook_iterator` should only impact `__iter__` of `IterDataPipe` since `MapDataPipe` has random access to the data and should not be constrained.

**Note that I plan to land this together the single iterator constraint PRs, so this will not land for now.**

Differential Revision: [D36156459](https://our.internmc.facebook.com/intern/diff/D36156459)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36156459).

NivekT committed:
Update on "[DataPipe] refactor to separate `_IterDataPipeMeta`"


There are some features of the metaclass that we would like to be unique to `IterDataPipe` and not applied to `MapDataPipe`. 

For example, `hook_iterator` should only impact `__iter__` of `IterDataPipe` since `MapDataPipe` has random access to the data and should not be constrained.

**Note that I plan to land this together the single iterator constraint PRs, so this will not land for now.**

Differential Revision: [D36156459](https://our.internmc.facebook.com/intern/diff/D36156459)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36156459).

NivekT committed:
Update on "[DataPipe] refactor to separate `_IterDataPipeMeta`"


There are some features of the metaclass that we would like to be unique to `IterDataPipe` and not applied to `MapDataPipe`. 

For example, `hook_iterator` should only impact `__iter__` of `IterDataPipe` since `MapDataPipe` has random access to the data and should not be constrained.

**Note that I plan to land this together the single iterator constraint PRs, so this will not land for now.**

Differential Revision: [D36156459](https://our.internmc.facebook.com/intern/diff/D36156459)

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] refactor to separate `_IterDataPipeMeta`"


There are some features of the metaclass that we would like to be unique to `IterDataPipe` and not applied to `MapDataPipe`. 

For example, `hook_iterator` should only impact `__iter__` of `IterDataPipe` since `MapDataPipe` has random access to the data and should not be constrained.

**Note that I plan to land this together the single iterator constraint PRs, so this will not land for now.**

Differential Revision: [D36156459](https://our.internmc.facebook.com/intern/diff/D36156459)

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] refactor to separate `_IterDataPipeMeta`"


There are some features of the metaclass that we would like to be unique to `IterDataPipe` and not applied to `MapDataPipe`. 

For example, `hook_iterator` should only impact `__iter__` of `IterDataPipe` since `MapDataPipe` has random access to the data and should not be constrained.

**Note that I plan to land this together the single iterator constraint PRs, so this will not land for now.**

Differential Revision: [D36156459](https://our.internmc.facebook.com/intern/diff/D36156459)

[ghstack-poisoned]

VitalyFedyunin commented:
Suggestion, start landing this one now.

NivekT commented:
@pytorchbot merge this please

NivekT committed:
Update on "[DataPipe] refactor to separate `_IterDataPipeMeta`"


There are some features of the metaclass that we would like to be unique to `IterDataPipe` and not applied to `MapDataPipe`. 

For example, `hook_iterator` should only impact `__iter__` of `IterDataPipe` since `MapDataPipe` has random access to the data and should not be constrained.

**Note that I plan to land this together the single iterator constraint PRs, so this will not land for now.**

Differential Revision: [D36156459](https://our.internmc.facebook.com/intern/diff/D36156459)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36156459).

NivekT commented:
@pytorchbot merge this please
